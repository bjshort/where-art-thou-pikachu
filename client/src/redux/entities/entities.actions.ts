import { EntityState } from './entities.reducer';

export const UPDATE_ENTITIES = 'ENTITIES/UPDATE_ENTITIES';

export function updateEntities(payload: EntityState) {
  return {
    type: UPDATE_ENTITIES,
    payload,
  } as const;
}

export type EntityActionTypes = ReturnType<typeof updateEntities>;
