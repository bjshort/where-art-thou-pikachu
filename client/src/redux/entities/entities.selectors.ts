import { AppState } from '../root.reducer';

export const getPokemonById = (id?: number) => (state: AppState) => {
  if (!id || !state.entities.pokemon) return undefined;
  return state.entities.pokemon[id];
};

/* 
  In production we would redesign the store so that pokemon could also be looked up by name without
  looping through the existing object. We would also have an object that had name keys and id values */
export const getPokemonByName = (name: string) => (state: AppState) => {
  return Object.values(state.entities.pokemon).find(
    (pokemon) => pokemon.name.toLowerCase() === name.toLowerCase(),
  );
};
