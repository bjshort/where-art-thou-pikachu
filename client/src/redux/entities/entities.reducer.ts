import { PokemonDTO } from '../../api-client/pokemon.dto';
import { EntityActionTypes, UPDATE_ENTITIES } from './entities.actions';
import { merge } from 'lodash';
import { combineReducers, AnyAction } from 'redux';

export interface EntityState {
  pokemon: { [key: string]: PokemonDTO };
}

export const initialState: EntityState = {
  pokemon: {},
};

export default (
  state: EntityState = initialState,
  action: EntityActionTypes,
) => {
  switch (action.type) {
    case UPDATE_ENTITIES:
      return merge({}, state, action.payload);
  }

  return combineReducers<EntityState>({
    pokemon: (state: any = {}, action: AnyAction) => state,
  })(state, action);
};
