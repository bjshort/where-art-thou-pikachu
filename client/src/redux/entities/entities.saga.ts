import { Schema, normalize } from 'normalizr';
import { put } from 'redux-saga/effects';

import { EntityActionTypes, updateEntities } from './entities.actions';

export function* syncEntitiesAndGetResults(data: any, schema: Schema<any>) {
  const { entities, result } = normalize(data, schema);

  yield put<EntityActionTypes>(updateEntities(entities));

  return result;
}
