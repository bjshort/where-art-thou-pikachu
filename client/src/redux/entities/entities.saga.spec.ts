import { runSaga } from '@redux-saga/core';
import { syncEntitiesAndGetResults } from './entities.saga';
import SCHEMA from '../schema';

describe('entities Sagas', () => {
  let dispatchedActions: any[];
  let mockState: any;
  let mockStore: any;

  beforeEach(() => {
    dispatchedActions = [];

    mockState = {};

    mockStore = {
      getState: () => mockState,
      dispatch: (action: any) => dispatchedActions.push(action),
    };
  });
  describe('syncEntitiesAndGetResults', () => {
    it('should dispatch correct actions and return a list of ids', async () => {
      const id = await runSaga(
        mockStore,
        syncEntitiesAndGetResults,
        {
          id: 143,
          name: 'snorlax',
          description: {
            original:
              'Very lazy. Just eats and sleeps. As its rotund bulk builds, it becomes steadily more slothful.',
            shakespeare:
              "Very distemperate. Just engluts and sleeps. As its rotund bulk builds,  't becomes steadily moo slothful.",
          },
          imageUrl:
            'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/143.png',
          height: '6.89',
          weight: '1014.11',
          exp: 189,
        },
        SCHEMA.pokemon,
      ).toPromise();
      expect(dispatchedActions).toMatchSnapshot();
      expect(id).toEqual(143);
    });
  });
});
