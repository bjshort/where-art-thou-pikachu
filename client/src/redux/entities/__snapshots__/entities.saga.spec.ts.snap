// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entities Sagas syncEntitiesAndGetResults should dispatch correct actions and return a list of ids 1`] = `
Array [
  Object {
    "payload": Object {
      "pokemon": Object {
        "143": Object {
          "description": Object {
            "original": "Very lazy. Just eats and sleeps. As its rotund bulk builds, it becomes steadily more slothful.",
            "shakespeare": "Very distemperate. Just engluts and sleeps. As its rotund bulk builds,  't becomes steadily moo slothful.",
          },
          "exp": 189,
          "height": "6.89",
          "id": 143,
          "imageUrl": "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/143.png",
          "name": "snorlax",
          "weight": "1014.11",
        },
      },
    },
    "type": "ENTITIES/UPDATE_ENTITIES",
  },
]
`;
